### PR TITLE
Remove StaticAssert* from Kokkos

### PR DESCRIPTION
**Pull request by ndellingwood**
Remove Kokkos_StaticAssert.hpp and remnant usage; replace with c++11
calls.
Addresses issue #2129

### CODE REVIEW

**Comment by ndellingwood** [comment]
All this stuff was in Impl namespace, so deprecation is not required.

File: core/perf_test/PerfTestBlasKernels.hpp [hunk 1]
```diff
@@ -124,14 +124,10 @@ struct Dot
   typedef typename Device::execution_space execution_space ;
 
   typedef typename
-    Impl::StaticAssertSame< Impl::unsigned_< 1 > ,
-                            Impl::unsigned_< Type::Rank > >::type ok_rank ;
+    std::is_same< Impl::unsigned_< 1 > ,
+                  Impl::unsigned_< Type::Rank > >::type ok_rank ;
```
Comment (dalg24) [inline core/perf_test/PerfTestBlasKernels.hpp]:
```C++
static_assert( std::is_same<A, B>::value );
```

and `#include <type_traits>`

Comment (mhoemmen) [reply core/perf_test/PerfTestBlasKernels.hpp]: Yeah, this shouldn't even compile.

File: core/perf_test/PerfTestBlasKernels.hpp [hunk 2]
```diff
@@ -165,12 +161,8 @@ struct DotSingle
   typedef typename Device::execution_space execution_space ;
 
   typedef typename
-    Impl::StaticAssertSame< Impl::unsigned_< 1 > ,
-                            Impl::unsigned_< Type::Rank > >::type ok_rank ;
-
-/*  typedef typename
-    Impl::StaticAssertSame< execution_space ,
-                            typename Type::execution_space >::type ok_device ;*/
+    std::is_same< Impl::unsigned_< 1 > ,
```
Comment (mhoemmen) [inline core/perf_test/PerfTestBlasKernels.hpp]: Ditto.

Comment (ndellingwood) [reply core/perf_test/PerfTestBlasKernels.hpp]: Fixed.

File: core/perf_test/PerfTestBlasKernels.hpp (hunk 1 above)
Comment (ndellingwood) [reply core/perf_test/PerfTestBlasKernels.hpp]: Fixed

**Comment by ndellingwood** [comment]
Tested perf_tests locally, compiled and ran fine.
Spot-check on kokkos-dev-2 passed except cuda - I forgot to set the arch for Volta, will rerun just with cuda.
Fixes for Stokhos on my local kokkos-promotion branch of Trilinos that I'll push there.

```
Running on machine: sems
Repository Status:  49147805be617910f5d146f0d5f469ac0debd5fe Remove commented line


Going to test compilers:  gcc/5.3.0 gcc/7.2.0 intel/17.0.1 clang/3.9.0 cuda/9.2
Testing compiler gcc/5.3.0
Testing compiler gcc/7.2.0
  Starting job gcc-5.3.0-OpenMP-release
  PASSED gcc-5.3.0-OpenMP-release
Testing compiler intel/17.0.1
  Starting job gcc-7.2.0-Serial-release
  PASSED gcc-7.2.0-Serial-release
Testing compiler clang/3.9.0
  Starting job intel-17.0.1-OpenMP-release
  PASSED intel-17.0.1-OpenMP-release
Testing compiler cuda/9.2
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
```

Rerunning cuda with proper arch flag.

File: core/perf_test/PerfTestBlasKernels.hpp (hunk 2 above)
Comment (dalg24) [reply core/perf_test/PerfTestBlasKernels.hpp]: Any reason `static_assert( Type::Rank == 1, "...");` would not work?

Comment (mhoemmen) [reply core/perf_test/PerfTestBlasKernels.hpp]: `Rank` has type `int` except for Stokhos View specializations.  I usually just cast `Type::Rank` to `int` and everything works fine.

**Comment by ndellingwood** [comment]
Looks like the failed build was a Travis issue. @crtrott @ibaned review comments addressed and spot-check passed, this should be ready for merge.

**Comment by mhoemmen** [comment]
@dalg24 We should actually get Stokhos or Sacado or both to fix the definition of Rank in their specializations of View, since it's an enum there and the compiler complains about `ViewType::Rank` comparisons with integers.

**Comment by dalg24** [comment]
> @dalg24 We should actually get Stokhos or Sacado or both to fix the definition of Rank in their specializations of View, since it's an enum there and the compiler complains about `ViewType::Rank` comparisons with integers.

Sure but either (or both) operand can be casted before the comparison.

**Comment by ndellingwood** [comment]
@dalg24 


> Please replace all the...

Updated.

File: core/perf_test/PerfTestBlasKernels.hpp [hunk 3]
```diff
@@ -125,9 +125,7 @@ struct Dot
 {
   typedef typename Device::execution_space execution_space ;
 
-  static_assert(
-    std::is_same< Impl::unsigned_< 1 > ,
-                  Impl::unsigned_< Type::Rank > >::value,
+  static_assert( static_cast<unsigned>(Type::Rank) == static_cast<unsigned>(1),
```
Comment (mhoemmen) [inline core/perf_test/PerfTestBlasKernels.hpp]: type of `1` is `int`.  Why not just cast the left-hand side to `int`?

**Pull request merged**